sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -218,7 +247,14 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       bucketSpec = getBucketSpec,
       options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    val destination = source match {
+      case "jdbc" => extraOptions.get(JDBCOptions.JDBC_TABLE_NAME)
+      case _ => extraOptions.get("path")
```

Comment:
> e.g. calling the save method adds a "path" key to the option map, but is that key name a public API?

yes, it is. e.g. `df.write.format("parquet").option("path", some_path).save()`, the `path` is a "magic key" and we've exposed it to users, so `path` is a public API and if we change it, we will break existing applications.